README
===============================================================================
 Link Monitor Applet overview
===============================================================================

Index

	0. About this document
	1. About Link Monitor Applet
	2. Dependencies
	3. Installation instructions

0. About this document

	$Id: README,v 1.8 2008-02-20 16:51:41 jylefort Exp $

	Copyright (C) 2004-2006 Jean-Yves Lefort.

	Permission is granted to copy, distribute and/or modify this
	document under the terms of the GNU Free Documentation License
	(GFDL), Version 1.2 or any later version published by the Free
	Software Foundation with no Invariant Sections, no Front-Cover
	Texts, and no Back-Cover Texts. You can find a copy of the
	GFDL at http://www.gnu.org/licenses/fdl.html or in the file
	COPYING-DOCS distributed with this manual.

1. About Link Monitor Applet

	Link Monitor Applet is a GNOME Panel Applet displaying the
	round-trip time to a set of hosts in a bar graph.

	Link Monitor Applet features include:

		* A line graph able to display up to one week of
		  round-trip time data
		* Country flags and names
		* Full ICMP and ICMPv6 support
		* HIG 2.0 compliance

2. Dependencies

	Mandatory:

		* GTK+ (version 2.12.0 or superior)
		  http://www.gtk.org/

		* GNOME (version 2.6.0 or superior)
		  http://www.gnome.org/

	Optional:

		* for country information support:

		  GeoIP
		  http://www.maxmind.com/geoip/

		  and a GTK+ SVG loader, for instance:

		  librsvg (version 2.0 or superior)
		  http://librsvg.sourceforge.net/

3. Installation instructions

	To configure, build and install Link Monitor Applet, use the
	following commands:

		./jb configure
		./jb build
		sudo ./jb install

	Note that a number of settings can be passed to the configure
	stage, for instance:

		./jb configure prefix=/opt cflags=-O3 geoip=no

	For details, use:

		./jb help

	For more informations about JB (the Jean-Yves Lefort's Build
	System), read jbsrc/lib/README.

	IMPORTANT:

		* The install phase will enable the setuid bit on the
		  link-monitor-applet binary (root privileges are
		  needed to create the ICMP and ICMPv6 sockets upon
		  applet startup; they will be dropped after that
		  step). You therefore need to run the install stage
		  as root, even for installing Link Monitor Applet in
		  your home directory.

		* Due to a bug in bonobo-activation
		  (http://bugzilla.gnome.org/show_bug.cgi?id=151082),
		  your session must be restarted after installing Link
		  Monitor Applet.


NEWS
3.0					February 20, 2008

	* The tooltip graphs have been merged into a single larger
	  graph. This new graph features many improvements, such as
	  the elimination of the staircase effect that was visible
	  with large echo delays, a configurable span, a better
	  appearance, and more.
	* The appearance of the applet bar graph has been improved
	* Host colors are now configurable
	* Time settings are now entered using sliders; these sliders
	  are non-linear and allow to easily choose an appropriate
	  value without having to use the keyboard
	* The reply timeout setting has been removed; the timeout is
	  now automatically computed from the maximum round-trip time
	* The minimum delay between echo requests has been lowered to
	  200 ms
	* ICMP/ICMPv6 error messages (destination unreachable, etc)
	  are now handled properly and displayed in the tooltip
	* ICMP/ICMPv6 packet sizes have been reduced to the smallest
	  possible value (no payload); they are 12 bytes smaller than
	  in the previous version
	* Incoming ICMP packets having an invalid checksum are now
	  ignored
	* A crash that occurred when the Ubuntu Human GTK+ theme was
	  in use has been worked around (thanks to Tomáš Hnyk for his
	  help); I believe that this crash was caused by a bug in the
	  Human theme engine
	* libeel is no longer used since its API is not stable
	* The build system has been changed from the GNU Autotools to
	  JB (the Jean-Yves Lefort's Build System)
	* The documentation is now licensed under the terms of the GNU
	  Free Documentation License version 1.2
	* Updated translation: French
	* New translation: Italian (Gabriele Vivinetto)

2.2					September 2, 2007

	* The applet icon is now displayed permanently
	* Could no longer close the about dialog with recent GTK+
	  versions: fixed
	* Using the host list buttons while editing a host caused a
	  crash: fixed
	* Some build warnings have been fixed
	* The src/automake.xml file is now distributed
	* Link Monitor Applet is now licensed under the terms of the
	  GNU General Public License version 3
	* The documentation is now licensed under the terms of the GNU
	  Free Documentation License
	* Updated translation: French
	* New translation: Bulgarian (Yavor Doganov)

2.1					June 22, 2006

	* ICMP replies were discarded if the ICMP sequence number
	  wrapped while the monitored host was down: fixed
	* Some build failures have been fixed

2.0					June 20, 2006

	* The applet tooltip has been improved: it now displays the
	  geographical location of hosts (country flag and name) as
	  well as round-trip time graphs
	* The panel background transparency/color is now respected
	* The applet is now activated with a single click; this is
	  consistent with HIG recommendations as well as other applets
	* Multiple applets can now be used
	* A crash which sometimes occurred on logout has been fixed
	* If the system resolver is not reentrant,
	  getaddrinfo()/getnameinfo() calls are now serialized
	* The ICMP/ICMPv6 payload size has been reduced in some cases
	* Some potential network byte order issues have been fixed
	* Replies received out of order are now handled properly
	* Link Monitor Applet is now licensed under the terms of the
	  GNU General Public License (previously, a BSD-style license
	  was used)
	* Updated translation: French

1.3					February 1, 2005

	* IP addresses are now resolved periodically
	* The host list can now be reordered using drag and drop
	* Now compiles cleanly even when -fstrict-aliasing is in use
	* The applet icon has been modified
	* Some cleanups have been performed

1.2					October 9, 2004

	* Binaries compiled with GCC 3.4.x were crashing in various
	  places: fixed
	* The manual is now installed in the proper location
	* Some cleanups have been performed
	* A few minor issues have been fixed

1.1					August 22, 2004

	* The preferences menu item was broken and has been fixed
	* The host list now has keyboard shortcuts
	* A few items have been added to the host list popup menu
	* Objects are now generated with GOB
	* A few minor issues have been fixed

1.0					August 15, 2004

	* A bug which caused the applet to hang on Linux (and on other
	  platforms having non-recursive mutexes) has been fixed
	* HIG compliance has been upgraded to version 2.0
	* Alt+Return now invokes the preferences

0.99					July 27, 2004

	* Initial release


jbsrc/jb.c
/*
 * Link Monitor Applet
 * Copyright (C) 2004-2008 Jean-Yves Lefort <jylefort@brutele.be>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <string.h>
#include "jb.h"

#jb_include "lm-gdk-pixbuf-loader"
#jb_include "lm-geoip"
#jb_include "lm-glibc"

#define LM_WARNING_CFLAGS \
  "-Wall "				\
  "-Wformat-y2k "			\
  "-Wformat-security "			\
  "-Wno-unused-parameter "		\
  "-Wdeclaration-after-statement "	\
  "-Wendif-labels "			\
  "-Wpointer-arith "			\
  "-Wcast-align	"			\
  "-Waggregate-return "			\
  "-Wmissing-noreturn "			\
  "-Wmissing-format-attribute "		\
  "-Wpacked "				\
  "-Wredundant-decls "			\
  "-Wnested-externs "			\
  "-Winline "				\
  "-Wno-pointer-sign "			\
  "-Wshadow "

static const JBFeature *jb_features[] = {
  &jb_pkg_config_feature,
  &jb_gettext_feature,
  &jb_intltool_feature,
  &jb_gconf_feature,
  &jb_gnome_help_feature,
  &jb_gob2_feature
};

void
jb_package_init (void)
{
  JBVariableGroup *lm_features_group;

  /* by default, use the GNOME prefix */
  jb_set_prefix_from_program("gnome-open");

  /*
   * At the time of this writing, gob2 2.0.16 is not yet
   * released. What is required is gob2 2.0.15 + my patches (see the
   * gob mailing list), which should eventually become gob2 2.0.16.
   */
  jb_variable_set_string("gob2-minversion", "2.0.16");

  lm_features_group = jb_variable_add_group("Link Monitor Applet features");

  jb_variable_add_bool("ipv6",
		       "enable IPv6 support",
		       lm_features_group,
		       JB_VARIABLE_C_DEFINE,
		       TRUE);
  jb_variable_add_bool("geoip",
		       "enable GeoIP support",
		       lm_features_group,
		       JB_VARIABLE_C_DEFINE,
		       TRUE);

  jb_variable_add_bool("compile-warnings", NULL, NULL, 0, FALSE);
  jb_variable_add_bool("debug", NULL, NULL, 0, TRUE);
  jb_variable_add_bool("suid-root", NULL, NULL, 0, TRUE);
}

static void
report_invalid_prefix (void)
{
  const char *gnome_prefix;

  jb_check_gnome_prefix();

  gnome_prefix = jb_variable_get_string("gnome-prefix");
  if (*gnome_prefix == '\0')
    jb_warning_expand("$human-package must be installed in the same prefix as GNOME, but the GNOME prefix could not be found. $human-package might be inoperable.", NULL);
  else
    {
      char *prefix;

      prefix = jb_variable_expand("$prefix", NULL);
      if (strcmp(prefix, gnome_prefix))
	jb_warning_expand("the $human-package prefix ($prefix) is different from the GNOME prefix ($gnome-prefix). $human-package will be inoperable. Unless you know what you are doing, please run \"./jb configure prefix=$gnome-prefix\".", NULL);
      g_free(prefix);
    }
}

void
jb_package_configure (void)
{
  if (lm_glibc_check())
    {
      /*
       * We need -std=c99 for lround(), ...
       * We need _BSD_SOURCE (which requires -lbsd-compat) for struct icmp, ...
       * We need _POSIX_C_SOURCE for struct addrinfo, clock_gettime(), ...
       */
      jb_variable_set_string("glibc-cflags", "-std=c99 -D_BSD_SOURCE -D_POSIX_C_SOURCE=199309L");
      jb_variable_set_string("glibc-cppflags", "");
      jb_variable_set_string("glibc-ldflags", "-lbsd-compat");
    }
  else
    {
      jb_variable_set_string("glibc-cflags", "");
      jb_variable_set_string("glibc-cppflags", "");
      jb_variable_set_string("glibc-ldflags", "");
    }

  if (jb_check_functions("clock_gettime", "rt"))
    jb_variable_set_string("clock-gettime-ldflags", "-lrt");
  else if (jb_check_functions("clock_gettime", NULL))
    jb_variable_set_string("clock-gettime-ldflags", "");
  else
    jb_error("clock_gettime() found neither in librt nor in libc");

  jb_check_reentrant_dns_resolver();

  jb_require_packages("GNOME", "gnome", "gthread-2.0 gtk+-2.0 >= 2.12 libgnomeui-2.0 libpanelapplet-2.0 libglade-2.0");

  if (jb_variable_get_bool("geoip"))
    {
      if (lm_geoip_check())
	{
	  if (! lm_gdk_pixbuf_loader_check("art/flags/fr.svg", "image/svg+xml"))
	    jb_warning("GTK+ cannot load image/svg+xml images, country flags will not be displayed");
	}
      else
	{
	  jb_warning("disabling option \"geoip\" since GeoIP was not found");
	  jb_variable_set_bool("geoip", FALSE);
	}
    }

  report_invalid_prefix();
}

void
jb_package_add_resources (void)
{
  JBGroup *group;
  JBObject *object;

  if (jb_variable_get_bool("compile-warnings"))
    jb_compile_options_add_cflags(&jb_compile_options, LM_WARNING_CFLAGS " -Werror");

  if (! jb_variable_get_bool("debug"))
    jb_compile_options_add_cflags(&jb_compile_options, "-DG_DISABLE_ASSERT -DG_DISABLE_CHECKS -DG_DISABLE_CAST_CHECKS");

  jb_compile_options_add_gob2flags(&jb_compile_options, "--exit-on-warn");

  /*** art *******************************************************************/

  group = jb_group_new("art");

  jb_group_add_data_file(group, "link-monitor-applet.png", "$datadir/pixmaps");

  if (jb_variable_get_bool("geoip"))
    jb_group_add_data_files_pattern(group, "flags/*.svg", "$pkgdatadir/flags");

  jb_group_add(group);

  /*** data ******************************************************************/

  group = jb_group_new("data");

  jb_group_add_resource(group, JB_GROUP_RESOURCE(jb_template_new("GNOME_LinkMonitorApplet.server.in.in")));
  jb_group_add_bonobo_server(group, "GNOME_LinkMonitorApplet.server.in");

  jb_group_add_gconf_schemas(group, "link-monitor-applet.schemas.in");

  jb_group_add(group);

  /*** help ******************************************************************/

  group = jb_group_new("help");

  jb_group_add_resource(group, JB_GROUP_RESOURCE(jb_gnome_help_new("C", "documentation-license.xml software-license.xml")));

  jb_group_add(group);

  /*** jbsrc *****************************************************************/

  group = jb_group_new("jbsrc");

  jb_group_add_dist_files(group,
			  "lm-gdk-pixbuf-loader-test.c",
			  "lm-glibc-test.c",
			  NULL);

  jb_group_add(group);

  /*** po ********************************************************************/

  group = jb_group_new("po");

  jb_group_add_translations(group, "bg fr it");

  jb_group_add(group);

  /*** src *******************************************************************/

  group = jb_group_new("src");

  object = JB_OBJECT(jb_program_new("link-monitor-applet"));

  jb_install_options_set_installdir(&object->install_options, "$libexecdir");

  if (jb_variable_get_bool("suid-root"))
    {
      jb_install_options_set_owner(&object->install_options, "root");
      jb_install_options_set_extra_mode(&object->install_options, S_ISUID);
    }

  jb_compile_options_add_string_defines(&object->compile_options,
					"PACKAGE", "$package", 
					"VERSION", "$version",
					"GETTEXT_PACKAGE", "$package",
					"PREFIX", "$prefix",
					"SYSCONFDIR", "$sysconfdir",
					"DATADIR", "$datadir",
					"PKGDATADIR", "$pkgdatadir",
					"LIBDIR", "$libdir",
					"GNOMELOCALEDIR", "$datadir/locale",
					"GNOMEPIXMAPSDIR", "$datadir/pixmaps",
					NULL);

  /*
   * We need --export-dynamic because because libglade needs to
   * resolve symbols from our own binary in order to autoconnect
   * signal handlers.
   */
  jb_compile_options_add_ldflags(&object->compile_options, "-Wl,--export-dynamic");
    
  jb_compile_options_add_ldflags(&object->compile_options, "-lm");

  jb_compile_options_add_package(&object->compile_options, "glibc");
  jb_compile_options_add_ldflags(&object->compile_options, "$clock-gettime-ldflags");
  jb_compile_options_add_package(&object->compile_options, "gettext");
  jb_compile_options_add_package(&object->compile_options, "gnome");

  if (jb_variable_get_bool("geoip"))
    jb_compile_options_add_package(&object->compile_options, "geoip");

  jb_object_add_sources(object,
			"LM:About:Dialog",
			"LM:Applet",
			"LM:Bar:Graph",
			"LM:Cell:Renderer:Color",
			"LM:Data:Set",
			"LM:Dialog",
			"LM:Framed:Image",
			"LM:Host",
			"LM:Host:Frontend",
			"LM:Host:View",
			"LM:Line:Graph",
			"LM:Line:Graph:Icon",
			"LM:Preferences:Dialog",
			"LM:Shell",
			"LM:Tooltips",
			"lm-color-generator",
			"lm-decls.h",
			"lm-icmp",
			"lm-main.c",
			"lm-non-linear-range",
			"lm-util",
			NULL);

  jb_group_add_resource(group, JB_GROUP_RESOURCE(object));

  jb_group_add(group);

  /*** ui ********************************************************************/

  group = jb_group_new("ui");

  jb_group_add_data_files(group,
			  "preferences-dialog.glade", "$pkgdatadir",
			  "GNOME_LinkMonitorApplet.xml", "$datadir/gnome-2.0/ui",
			  NULL);

  jb_group_add(group);
}

JB_MAIN("link-monitor-applet", "3.0", "Link Monitor Applet")
